### PR TITLE
Use build time component interface

### PR DESCRIPTION
**Pull request by cibernox**
> Make it work, make it right, make it fast.  In that order.
> -- Kent Beck 

This is the second step.

### CODE REVIEW

File: lib/ast-transform.js [hunk 1]
```diff
@@ -68,9 +68,125 @@
...
+    let iconValue;
+    let pair = this.node.hash.pairs.find((p) => p.key === 'icon');
+    if (pair) {
+      iconValue = pair.value;
+    } else {
+      iconValue = this.node.params[0];
+    }
+    if (iconValue) {
+      if (isDynamic(iconValue)) {
+        this.options.addon.fontAwesomeUsage.usedIconsUnknown = true;
```
Comment (jrjohnson) [inline lib/ast-transform.js]: I think `usedIconsUnknown` would make more sense as `usedDynamicIcons` or `dynamicIconsUsed`.

I'm also looking forward at FA5 which has some icon changes and we may want to create an `unknown` icon list to print any issues.

File: lib/ast-transform.js [hunk 2]
```diff
@@ -68,9 +68,125 @@
...
+      }
+    } else {
+      return parts.map((p) => p === '$' ? pair.value.value : p).join('');
+    }
+  } else if (component.options[propName]) {
+    return parts.map((p) => p === '$' ? component.options[propName] : p).join('');
+  }
+}
+
+class FaIconComponent extends BuildTimeComponent {
```
Comment (jrjohnson) [inline lib/ast-transform.js]: I think it would make sense to separate this class into its own file for clarity.

Comment (cibernox) [reply lib/ast-transform.js]: It does make sense, but while I iron the kinds I prefer to have it on the same file. I'll extract the to individual files at the very end.

**Comment by cibernox** [comment]
@jrjohnson This is ready for a review. 
The `BuiltTimeComponent` abstraction I've been working is pretty nice to use already, and makes creating components reasonably readable.

Per example, the component that transforms 
```hbs
{{#fa-list class="my-list" as |l|}}
  {{l.fa-icon "credit-card"}}
{{/fa-list}}
```
into
```html
<ul class="fa-ul my-list">
  <i class="fa fa-credit-card fa-li"></i>
</ul>
```
is just
```js
class FaListComponent extends BuildTimeComponent {
  constructor(node, opts = {}) {
    super(node, Object.assign({ tagName: 'ul', classNames: ['fa-ul'] }, opts));
    this.contentVisitor = {
      MustacheStatement: (node) => {
        let componentAlias = this.node.program.blockParams[0];
        if (componentAlias !== undefined && node.path.original === `${componentAlias}.fa-icon`) {
          return (new FaIconComponent(node, { listItem: true, addon: this.options.addon })).toElement();
        }
      }
    }
  }
}
```

The API is also designed for the future class properties (now in stage 3) and it will end up looking
```js
class FaListComponent extends BuildTimeComponent {
  tagName = 'ul'
  classNames = ['fa-ul']
  contentVisitor = {
    MustacheStatement: (node) => {
      let componentAlias = this.node.program.blockParams[0];
      if (componentAlias !== undefined && node.path.original === `${componentAlias}.fa-icon`) {
        return (new FaIconComponent(node, { listItem: true, addon: this.options.addon })).toElement();
      }
    }
  }
}
```

I probably will refine it further, but the base is set.

File: lib/build-time-components/fa-icon.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  30. <i aria-hidden="true" class="fa {{if boundValue (concat 'fa-' boundValue)}}"></i>
+  31. <i aria-hidden="true" class="fa {{if boundValue (concat 'fa-' boundValue)}} fa-fw"></i>
+  31. <i aria-hidden="true" class="fa {{if boundValue (concat 'fa-' boundValue)}} fa-pull-{{direction}}"></i>
+  ```
+*/
+const { builders: b } = require('@glimmer/syntax');
+const { BuildTimeComponent } = require('ember-ast-helpers');
+const { isDynamic, buildConcatIfPresent, interpolateSize } = require('./utils');
+
+function interpolateAttribute(component, propName, parts) {
```
Comment (jrjohnson) [inline lib/build-time-components/fa-icon.js]: Any reason `interpolateSize` is imported and `interpolateAttribute` is here?

File: testem.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ module.exports = {
   test_page: 'tests/index.html?hidepassed',
   disable_watching: true,
   launch_in_ci: [
-    'Firefox',
+    // 'Firefox',
```
Comment (jrjohnson) [inline testem.js]: Not sure why you're disabling this here, but it should be removed or uncommented. One or the other.

**Comment by jrjohnson** [comment]
I may not have time to test this weekend, but I will try and get to it ASAP. I stuck some quick code review comments in.

I really like the API for the classes themselves. I think it breaks down fairly readably and I was able to understand what most of the code does just reading it which is pretty sweet! 

One that that jumps out at me is that I'm not sure how `toElement()` knows to call `iconContent` when building. I looked over the docs for `ember-ast-helpers` and I see details about `propName` and `<propName>Content()` but I don't see where `propName` is coming from in these classes as it doesn't seem to be getting declared.

File: lib/build-time-components/fa-icon.js (hunk 3 above)
Comment (cibernox) [reply lib/build-time-components/fa-icon.js]: I was planning to extract `interpolateAttribute` to `ember-ast-helpers`, as it's generally useful.
Which I did now.

 The interpolate size specific for this project.

**Comment by cibernox** [comment]
@jrjohnson Actually is quite simple. The `icon` is `classNameBinding` of this project, and the value of any binding is the return value of the `<propName>Content` (if  defined).

So you see it clear:
1) All BuildTimeComponents have an `attributeBinding` for class: https://github.com/cibernox/ember-ast-helpers/blob/master/lib/build-time-component.ts#L76
2) All `BuildTimeComponent` have defined `classContent` function that classNames, classNameBindings and the `class=whatever` attribute passed in the template: https://github.com/cibernox/ember-ast-helpers/blob/master/lib/build-time-component.ts#L138
3) `FaIcon` defines a classNameBinding for `icon`: https://github.com/martndemus/ember-font-awesome/pull/152/files#diff-f4f5cf78e6c12b18d6f0449057dfa288R81
4) Every classNameBinding attempts to use `<propName>` as the first source for its content. It not present, it will use `propName="somthing"` from the template. If that also fails, it will use the default values passed on `new MyComponent(node, { propName: 'static value' })`: https://github.com/cibernox/ember-ast-helpers/blob/master/lib/build-time-component.ts#L284

**Comment by cibernox** [comment]
I believe all feedback has been addressed. Merging.

**Pull request merged**